File: solutions/ruby/matching-brackets/1/matching_brackets.rb
module Brackets
  def self.paired?(string)
    return true if string.empty?
    parens = []
    string.chars.each do |char|
      if char == '{' || char == '[' || char == '('
        parens.push(char)
      else 
        if char == '}'
          parens.last == '{' ? parens.pop : parens.push(char)
        elsif char == ']'
          parens.last == '[' ? parens.pop : parens.push(char)
        elsif char == ')'
          parens.last == '(' ? parens.pop : parens.push(char)
        end 
      end
    end 
    parens.empty? 
  end
end 

File: solutions/cpp/matching-brackets/1/matching_brackets.cpp
#include "matching_brackets.h"
#include <stack> 
namespace matching_brackets {
    bool check(const string& s) {
        stack<char> brackets;

        for (const char c : s) {
            switch(c) {
                case '(': 
                case '[':
                case '{':
                    brackets.push(c);
                    break;
                case ')':
                    if (brackets.empty() || brackets.top() != '(' ){
                        return false;
                    }
                    brackets.pop();
                    break;
                case ']':
                    if (brackets.empty() || brackets.top() != '[' ){
                        return false;
                    }
                    brackets.pop();
                    break;
                case '}':
                    if (brackets.empty() || brackets.top() != '{' ){
                        return false;
                    }
                    brackets.pop();
                    break; 
                default:
                    break; 
            }
        }

        return brackets.empty();
    }
}  // namespace matching_brackets
